packages/crayons-core/src/components/file-uploader/file-uploader.tsx

```diff
@@ -0,0 +1,524 @@
+   */
+  @Prop() maxFileSize = 0;
+
+  /**
+   * acceptError - Error message to display when format is invalid.
+   */
+  @Prop() acceptError = 'File format not accepted';
+
+  /**
+   * maxFileSizeError - Error message to display when file size exceeds limit
+   */
+  @Prop() maxFileSizeError = 'Exceeded maximum file size';
+
+  /**
+   * actionURL - URL to make server call.
+   */
+  @Prop() actionURL = '';
+
+  /**
+   * actionParams - additional information to send to server other than the file.
+   */
+  @Prop() actionParams: any = {};
+
+  /**
+   * multiple - upload multiple files.
+   */
+  @Prop() multiple = false;
+
+  /**
+   * modify request
+   * @param xhr
+   * @returns xhr
+   */
+  @Prop() modifyRequest: (xhr: any) => any = (xhr) => xhr;
+
+  /**
+   * files - files collection.
+   */
+  @State() files: any = [];
+
+  /**
+   * errors - errors collection.
+   */
+  @State() errors: any = [];
+
+  /**
+   * filesUploaded - event that gets emitted when files get uploaded
+   */
+  @Event() filesUploaded: EventEmitter;
+
+  /**
+   * fileReuploaded - event that gets emitted when file is reuploaded
+   */
+  @Event() fileReuploaded: EventEmitter;
+
+  /**
+   * stageChanged - event that gets emitted when component stage changes
+   */
+  @Event() stageChanged: EventEmitter;
+
+  /**
+   * private
+   * fileInputElement
+   */
+  fileInputElement: HTMLElement = null;
+
+  /**
+   * private
+   * isFileUploadInProgress
+   */
+  isFileUploadInProgress = false;
+
+  /**
+   * private
+   * fileUploadPromises
+   */
+  fileUploadPromises: any = [];
+
+  /**
+   * private
+   * formDataCollection
+   */
+  formDataCollection: any = {};
+
+  @Watch('stage')
+  stageChange(newStage) {
+    switch (newStage) {
+      case 'dropzone':
+        this.formDataCollection = {};
+        this.fileUploadPromises = [];
+        this.errors = [];
+        this.files = [];
+        break;
+      default:
+        break;
+    }
+    this.stageChanged.emit(newStage);
+  }
+
+  /**
+   * private
+   * uploadFileLocally - upload the files locally and add it to form for sending to server
+   * @param file
+   */
+  uploadFileLocally(file) {
+    const formData = new FormData();
+    formData.append('file', file);
+    this.formDataCollection[fileCount] = formData;
+    this.files.push({
+      id: fileCount,
+      name: file.name,
+      progress: 0,
+      error: '',
+    });
+    fileCount = fileCount + 1;
+  }
+
+  /**
+   * uploadFile
+   * @param fileId
+   * @returns fileUploadPromise
+   */
+  uploadFile(fileId) {
+    const formData = this.formDataCollection[fileId];
+    // adding extra information to formData before uploading
+    for (const key in this.actionParams) {
+      if (Object.prototype.hasOwnProperty.call(this.actionParams, key)) {
+        formData.append(key, this.actionParams[key]);
+      }
+    }
+    // creating and sending xhr requests
+    const xhr = new XMLHttpRequest();
```

Comment:
Kaushal, considered both options - making server call from component, delegating to customer. Since a lot of UI state of this component is based on the server call, decided to keep it in the component. Referred to couple of other libraries and they have used the same strategy.

As we are giving option to modify the xhr request / the payload before sending a request, we can customise this request to any requirement. We can use the 'modifyRequest' prop for adding headers. We can use 'actionParams' prop to send additional information in the formData. 

Tested these on the CO case, this is working fine.